more flexible type of input arguments for px functions

emmanuelle opened the pull request:
This is a first a draft for accepting more flexible types of input arguments for `px` functions, eg numpy arrays, dataframe columns, mixed type etc. At the moment the transformation done by `build_or_augment_dataframe` is done argument by argument, so having different arguments with different types is ok. 

Todo:
- write more tests
- decide what to do with arguments which are lists, such as `hover_data`, `custom_data`, `dimensions`. At the moment they are just skipped so they should be column names. I did this because I wanted to discuss the names they should have, should it be `customdata_0`, `customdata_1` etc. ?

For the cases which are taken into account, please see the test file.

Closes #1767 , follow-up on https://github.com/plotly/plotly_express/pull/87 (thank you @malmaud !)

emmanuelle commented:
One other thing to fix : `px.scatter(tips, x="bla", y='wrong')` does not produce a meaningful error message any more, have to fix this.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,42 @@ def apply_default_cascade(args):
+    argument; for example `scatter(x=[1,2], y=[3,4])`.
+    """
+    if args.get("data_frame") is None:
+        df = pd.DataFrame()
```
arguably we could just always start with an empty DF and progressively add columns to it.

emmanuelle replied:
done

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,42 @@ def apply_default_cascade(args):
+        df = pd.DataFrame()
+    else:
+        df = args["data_frame"]
+        df = df.reset_index()
```
I don't think we should automatically do this, as it will flow through to the `dimension`-supporting traces :)

emmanuelle replied:
yes, this has been corrected

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import pandas as pd
+
+
```
here I think we can also unit-test the `build_or_augment_dataframe` function directly... probably easier to read the tests that way IMO

emmanuelle replied:
we can do both, it's also nice to have regression tests calling the px functions (which are here mostly to be sure it does not crash :-))

emmanuelle commented:
TODO
Additional cases to take into account: check that error messages make sense when:
- having different lengths for different kw arguments (or dimensions not compatible with the given DataFrame)
- what happens with 2D arrays passed as parameters

At the moment the initial DataFrame is modified in-place, probably we should not do that (unless the DataFrame is very big??).

emmanuelle commented:
next step: start from empty dataframe and add columns

nicolaskruchten commented:
We should also move the "non-existent column" check and error message into this new function, which I think will help with some of this logic.

emmanuelle commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,75 @@ def apply_default_cascade(args):
+    `attrables` is a list of keys into `args`, all of whose corresponding
+    values are converted into columns of a dataframe.
+    Used to be support calls to plotting function that elide a dataframe
+    argument; for example `scatter(x=[1,2], y=[3,4])`.
```
need to write docstring here

emmanuelle commented:
@nicolaskruchten this is ready for a round of review. In the meantime I'll write a better docstring for the helper function.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,90 @@ def apply_default_cascade(args):
+    # implicitely need all dimensions: Splom, Parcats, Parcoords
+    # This could be refined when dimensions is given
+    if constructor in [go.Splom, go.Parcats, go.Parcoords]:  # we take all dimensions
+        df = args["data_frame"]
```
even in this case I would not directly use the provided data frame: rather I would start with an empty `df` always, and move the default `dimensions` logic here (basically if `args["dimensions"]` is present but empty then add all the columns from `args["data_frame"]` to `df`)... using this logic would also enable you to not pass `constructor` in here. (In general, the less we can condition things on `constructor` the better IMO :)

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,90 @@ def apply_default_cascade(args):
+        df = pd.DataFrame()
+
+    # Retrieve labels (to change column names)
+    labels = args.get("labels")  # labels or None
```
is it important to move the (re-)labelling logic into this function? Right now it all mostly happens later with various helper functions. If we're moving this logic into this function then should we refactor it out of where it is right now...?

emmanuelle replied:
yes, it's better to leave the labelling logic later as it was before, this should be resolved now.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,90 @@ def apply_default_cascade(args):
+            # Case of numpy array or df column
+            else:
+                try:
+                    col_name = argument.name  # pandas df
```
what case does this cover?

nicolaskruchten replied:
oh I see now... This is potentially very dangerous.

nicolaskruchten replied:
We should only do this iff `argument` is not a column (as checked by `is`) *and* if `argument.name` doesn't appear in `df.columns`. That way we can accept `Series` from other dataframes or xarrays, say, but we don't grab `df.x` by name if the user passes in `df.x*10` (which, annoyingly, passes `df.x.name` through).

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,90 @@ def apply_default_cascade(args):
+                    df.insert(0, col_name, argument)
+                except ValueError:  # if col named index already exists, replace
+                    df[col_name] = argument
+            # Case of numpy array or df column
```
I'm not sure how/when this covers the case of `df.col` ? We don't seem to actually look to see `if argument in args["data_frame"]` the way I would expect... ?

nicolaskruchten replied:
here we should loop on columns by name and check for `is` equality on `argument`.

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,144 @@
+    assert fig.data[0]["hovertemplate"] == "index=%{x}<br>total_bill=%{y}"
+    # I was not expecting this to work but it does...
+    fig = px.scatter(tips, x=tips.index, y=10 * tips.total_bill)
+    assert fig.data[0]["hovertemplate"] == "index=%{x}<br>total_bill=%{y}"
```
wow, I'm surprised that this works ... how does it work with the `total_bill` name??

nicolaskruchten replied:
Oh I see. scary :)

emmanuelle replied:
For the docstrings of `px` functions I chose to use the name `array_like` which is commonly used in numpy to refer to both arrays and lists, and in the broad sense dataframes as well. This is hopefully ready for another round of review.

emmanuelle replied:
once black is happy of course...

emmanuelle commented:
For the docstrings of px functions I chose to use the name array_like which is commonly used in numpy to refer to both arrays and lists, and in the broad sense dataframes as well. This is hopefully ready for another round of review.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,107 @@ def apply_default_cascade(args):
+            else args.get(field_name)
+        )
+        if argument_list is None:  # argument not specified, continue
+            continue
```
what if it's `[None]` ?

nicolaskruchten replied:
I would maybe break out of the loop even earlier if `args.get(field_name) is None`

emmanuelle replied:
ok, taken into account

nicolaskruchten replied:
I think I would move this up to the top of the `for` loop as `if args.get(field_name) is None: continue` that way we don't need to check for that `[None]` case at all.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,107 @@ def apply_default_cascade(args):
+                continue
+            elif isinstance(argument, str):  # just a column name
+                # Check validity of column name
+                try:
```
what if `args["data_frame"]` isn't provided? we need a special error message for "you can't just give us strings without a dataframe!" or something

emmanuelle replied:
done!

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -821,6 +827,11 @@ def build_or_augment_dataframe(args, attrables, array_attrables):
                         "Expected one of %s but received: %s"
                         % (field, str(list(df_columns)), argument)
                     )
+                except TypeError:
```
This seems overly-broad to me, compared to just directly checking if the `data_frame` argument has been passed in... I worry about try-catch control flow as you never know what new crazy thing will raise any given exception later :)

emmanuelle replied:
ok, I hesitated indeed between the two solutions. I chose the try/except mostly for the sake of compactness since this function is getting crazily long but this is not such a good argument compared to robustness.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,119 @@ def apply_default_cascade(args):
+    )
+    group_attrs = ["symbol", "line_dash"]
+    for group_attr in group_attrs:
+        if group_attr in args:
```
this case is already handled below, so you may as well just add `symbol` and `line_dash` in attrables always at call-time?

emmanuelle replied:
ok done

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,119 @@ def apply_default_cascade(args):
+                        # we do not want to keep the name, revert to field
+                        col_name = (
+                            col_name
+                            if args["data_frame"][col_name].equals(argument)
```
what's the tradeoff between using `.equals()` vs `is` here? the former is slower but a bit more permissive in that if it's a copy with the same name we accept it?

emmanuelle replied:
yes this was my idea. But maybe the flexibility gain is not worth the performance cost, I'm also happy to change to `is`.

nicolaskruchten replied:
Maybe we can switch to `is` for maximum strictness/performance, to start, and see if people complain/notice?

emmanuelle replied:
done

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,112 @@ def apply_default_cascade(args):
+            elif isinstance(argument, str):  # just a column name
+                if not isinstance(args.get("data_frame"), pd.DataFrame):
+                    raise ValueError(
+                        "String arguments are only possible when a DataFrame"
```
missing a trailing space here: `DataFrameis`

nicolaskruchten commented:
So this gives an error, as it should, due to misaligned lengths: `px.scatter(px.data.tips(), x=[1,2,3,4], y=[1,2,3])`

But this does not: `px.scatter(px.data.tips(), x=[1,2,3,4], y="tip")`

emmanuelle commented:
Oh wow good catch. `pandas` accepts very weird things like 
```
In [10]: df = pd.DataFrame({'a':[1, 3], 'b':[5, 6], 'c':[1, 2]})        
In [11]: df['bla'] = tips['tip']    
```
does not raise an error

nicolaskruchten commented:
So this does something fun: `px.bar(x=[['a','b'],['a','c'],['d','b'],['d','c']], y=[1,2,3,4])` ... it triggers our "multicategory" behaviour, but not a in good way:

![image](https://user-images.githubusercontent.com/203523/65073141-3de4bf80-d960-11e9-8f0e-ddc2805f4a38.png)

nicolaskruchten commented:
so maybe we need to check on "shape" as well as just length?

emmanuelle commented:
I'm not sure it's possible to support multicategory axis if all values have to be in a dataframe at some point. For your example to work (like it does with `go`) we should have I think `x.bar(x=[['a','a', 'b', 'b'],[1, 2, 1, 2], y=[1,2,3,4]) ` and then because of dimension mismatch (we would need a transpose) it's not possible to put x and y in a common dataframe.

emmanuelle commented:
but the correct behavior is not right, that's for sure

emmanuelle commented:
Well the current behavior is not right because of what `go` does:
```
import plotly.graph_objects as go
x=[['a','b'],['a','c'],['d','b'],['d','c']]
y=[1,2,3,4]
fig = go.Figure(data=go.Bar(x=x, y=y)
)
fig.show()
```
does the same.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,138 @@ def apply_default_cascade(args):
+            df[df_args.columns] = df_args[df_args.columns]
+
+    # Cast data_frame argument to DataFrame (it could be a numpy array, dict etc.)
+    if args["data_frame"] is not None and not isinstance(
```
should this move up above the `dimensions` block, which relies on data frames a bit?

emmanuelle commented:
@jonmmease we've made some progress here and your thoughts would be very welcome, since it's a quite important change (more flexibility means empowering users, but also potentially having more bugs :-)). The test file should give you an idea of cases which are supported with this PR. What do you think of this new API ? Any changes you would suggest?

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,135 @@ def apply_default_cascade(args):
+                            if argument is args["data_frame"][col_name]
+                            else field
+                        )
+                except AttributeError:  # numpy array, list...
```
this is intended to catch errors from `argument.name` primarily, right? can we just check for the presence of this directly and avoid the try-catch pattern?

emmanuelle replied:
ok

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,135 @@ def apply_default_cascade(args):
+                continue
+            # Case of index
+            elif isinstance(argument, pd.core.indexes.multi.MultiIndex):
+                raise TypeError("pandas MultiIndex is not supported by plotly express")
```
for all these errors we should clearly indicate which attribute is causing the issue.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,135 @@ def apply_default_cascade(args):
+                except AttributeError:  # numpy array, list...
+                    col_name = field
+                if length and len(argument) != length:
+                    raise ValueError("All arguments should have the same length.")
```
we should indicate which attribute is causing the issue here (I know in case of length it's ambiguous, but any hint helps!)

nicolaskruchten replied:
maybe report the previous argument whose length we compared to?

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,135 @@ def apply_default_cascade(args):
+                    args.get("data_frame"), pd.DataFrame
+                ) and not isinstance(args.get("data_frame"), np.ndarray):
+                    raise ValueError(
+                        "String or int arguments are only possible when a"
```
we should indicate which arg this is

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,135 @@ def apply_default_cascade(args):
+            ):  # just a column name
+                if not isinstance(
+                    args.get("data_frame"), pd.DataFrame
+                ) and not isinstance(args.get("data_frame"), np.ndarray):
```
we no longer need to check against `ndarray` here right?

emmanuelle replied:
good catch!

nicolaskruchten commented on packages/python/plotly/plotly/express/_doc.py:
```diff
@@ -15,98 +17,110 @@
     data_frame=["A 'tidy' `pandas.DataFrame`"],
```
this line will likely need to change also :)

nicolaskruchten commented on packages/python/plotly/plotly/express/_doc.py:
```diff
@@ -1,7 +1,9 @@
 
-colref = "(string: name of column in `data_frame`)"
-colref_list = "(list of string: names of columns in `data_frame`)"
+colref = "(string: name of column in `data_frame`, or array_like object)"
```
should we add something about the fact that array_like objects with `.name` defined get special treatment? maybe not worth the bloat.

emmanuelle commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,169 @@ def apply_default_cascade(args):
+    elif field_name + argument not in df.columns:
+        return field_name + "_" + argument
+    else:
+        raise NameError("A name conflict was encountered.")
```
I will improve the error message here, I pushed just to show the logics.

emmanuelle commented:
Our name setting heuristic is interferring with `hover_data` and `custom_data`, will look into this

emmanuelle commented:
OK, with this name-changing heuristic it will be very complicated if we accept too many things. What I suggest is that
- if a dataframe is provided as `data_frame` then we get all string arguments as forbidden names, and we will change the name of iterables if they have a name in this list of forbidden names.
- if several iterables have the same name we can either have a "first come - first serve" policy but it's not great, or we could change all names like "df1_x" and "df2_x".

emmanuelle commented:
Now we have a better naming heuristic but the code is very ugly, please don't review now as I need to improve this

emmanuelle commented:
The `build_dataframe` function is now very long, mostly because of the necessity to handle a number of different cases for column names. This part of the code does not smell good, to the extent that I'm wondering whether we should not support less fancy things, like
- if the `data_frame` argument is provided, then arguments with str column names or column names from `df` (as tested with `is` and not `equals`) keep their names in df
- if `data_frame` is not provided, all arguments have kw names (`x`, `y`, `size`)
- if you want more control, you use the `labels` argument. 

This would be a basic but robust behaviour. Conversely, with the current behaviour we do some magics in order to use column names as much as possible. This is useful when several dataframes are used in the arguments, but it should be a marginal case I guess. 

With both solutions all `px` calls in the tests will work ie display the same data, up to hover data and axis labels.

To be discussed! @nicolaskruchten @jonmmease

nicolaskruchten commented:
Yeah, this is closer to the original goal here... It's been worthwhile exploring the idea of using `.name` from arbitrary vectors but in the details it does seem like the main source of issues IMO.

emmanuelle commented:
@nicolaskruchten ready for a round of review when you find the time. One thing left to discuss if the name that should be used when the argument is not a column from the `data_frame` argument. In the example below
```
import plotly.express as px
import pandas as pd
df = pd.DataFrame(dict(a=[0, 1], b=[1, 10], c=[0.1, 0.8]))
df2 = pd.DataFrame(dict(time=[23, 26], money=[100, 200]))
fig = px.scatter(df, x=df.a, y=df2.money, size=df.b)
```
what should be the name used for the `y` argument? At the moment it's just `y`, but we could also use `y_money`, ie concatenate the name of the keyword and the name of the column. This would maybe be more informative. At the moment `y_money` is only used if there is already a column `y` of `data_frame` which is used in the plot.

nicolaskruchten commented:
> At the moment y_money is only used if there is already a column y of data_frame which is used in the plot

This is fine. What happens if `data_frame` also contains `y_money`? :)

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,202 @@ def apply_default_cascade(args):
+                continue
+            if isinstance(arg, str):
+                used_col_names.add(arg)
+            if isinstance(arg, pd.DataFrame) or isinstance(arg, pd.core.series.Series):
```
we accept full `DataFrame`s as well as `Series`?

emmanuelle replied:
hum now, it won't work with a full `DataFrame` actually (an error will be raised because it has no `name`). I will correct the code and the doc.

emmanuelle replied:
ok, corrected

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,202 @@ def apply_default_cascade(args):
+        for arg in names:
+            if arg is None:
+                continue
+            if isinstance(arg, str):
```
what if `arg` isn't in `df` ?

emmanuelle replied:
it will be caught later on, L882 
```
 if argument not in args["data_frame"].columns:
                    raise ValueError(
                        "Value of '%s' is not the name of a column in 'data_frame'. "
                        "Expected one of %s but received: %s"
                        % (field, str(list(args["data_frame"].columns)), argument)
                    )
```
is it ok for you?

nicolaskruchten replied:
yes, I was just curious about why it ends up in the reserved names

emmanuelle replied:
we could also raise the error in the function building the list of reserved names, which is always called

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,202 @@ def apply_default_cascade(args):
+                arg_name = arg.name
+                if arg_name and hasattr(df, arg_name):
+                    in_df = arg is df[arg_name]
+                    if in_df:
```
can probably inline this check

nicolaskruchten replied:
oh I just meant `if arg is df[arg_name]` ... doesn't matter much :)

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,202 @@ def apply_default_cascade(args):
+                    raise ValueError(
+                        "All arguments should have the same length."
+                        "The length of column argument `df[%s]` is %d, whereas the"
+                        "length of previous arguments is %d"
```
is it possible to name the previous argument here? "previous" won't mean anything to the user as they don't know the iteration order

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,207 @@ def apply_default_cascade(args):
+            elif isinstance(argument, pd.core.indexes.multi.MultiIndex):
+                raise TypeError(
+                    "Argument '%s' is a pandas MultiIndex."
+                    "pandas MultiIndex is not supported by plotly express" % field
```
maybe add "at the moment" ? :)

emmanuelle replied:
done

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,207 @@ def apply_default_cascade(args):
+                    if not df_provided:
+                        col_name = field
+                    else:
+                        keep_name = hasattr(
```
instead of `hasattr`/`getattr` we could do `col_name in args["data_frame"].columns`/`args["data_frame"][col_name]` or something right?

nicolaskruchten commented on packages/python/plotly/plotly/express/_doc.py:
```diff
@@ -1,7 +1,9 @@
 
-colref = "(string: name of column in `data_frame`)"
-colref_list = "(list of string: names of columns in `data_frame`)"
+colref = "(string: name of column in `data_frame`, or array_like object)"
```
worth adding that `int` column names are accepted also?

emmanuelle replied:
done

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+
+
+def test_numpy():
+    fig = px.scatter(x=[1, 2, 3], y=[2, 3, 4], color=[1, 3, 9])
```
can we assert something here? :)

emmanuelle replied:
done

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+    df = pd.DataFrame(dict(x=[0, 1], y=[1, 10], z=[0.1, 0.8]))
+    df2 = pd.DataFrame(dict(time=[23, 26], money=[100, 200]))
+    fig = px.scatter(df, x="z", y=df2.money, size="y")
+    assert fig.data[0].hovertemplate == "z=%{x}<br>y_money=%{y}<br>y=%{marker.size}"
```
this result surprises me a bit...

emmanuelle replied:
why? It's a bit weird but it corresponds to the logics
- use the field name if possible
- if not use field + _ + name
Hopefully users will use sensible column names most of the time...

emmanuelle replied:
this has been chnged, it would raise an error

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+    df = pd.DataFrame(dict(x=[0, 1], y=[3, 4]))
+    fig = px.scatter(df, x=[10, 1], y="y", color="x")
+    assert np.all(fig.data[0].x == np.array([10, 1]))
+    assert fig.data[0].hovertemplate == "x_x=%{x}<br>y=%{y}<br>x=%{marker.color}"
```
`x_x` seems very surprising here also :)

emmanuelle replied:
would raise an error

emmanuelle commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+    with pytest.raises(ValueError) as err_msg:
+        fig = px.scatter(tips, x="index", y="total_bill")
+        assert (
+            "ValueError: Value of 'x' is not the name of a column in 'data_frame'"
```
we could, but this is quite a specific case. Would you like to print this only when the str is 'index'?

nicolaskruchten replied:
Yeah, that's a nice touch I think.

emmanuelle replied:
done!

emmanuelle commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+
+def test_pandas_series():
+    tips = px.data.tips()
+    before_tip = tips.total_bill - tips.tip
```
yes I will remove one of those

nicolaskruchten commented on packages/python/plotly/plotly/tests/test_core/test_px/test_px_input.py:
```diff
@@ -0,0 +1,267 @@
+        x="total_bill",
+        y="tip",
+        hover_data=[np.random.random(tips.shape[0])],
+        labels={"hover_data_0": "suppl"},
```
this is nice

emmanuelle replied:
and it should also go in the tutorial

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,207 @@ def apply_default_cascade(args):
+                df[col_name] = args["data_frame"][argument]
+                continue
+            # Case of multiindex
+            elif isinstance(argument, pd.core.indexes.multi.MultiIndex):
```
yep.

emmanuelle commented:
@nicolaskruchten I think I addressed all your comments. If you have a quick idea about the chart-studio related CI problem I'm all ears, otherwise I'll investigate elsewhere

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,215 @@ def apply_default_cascade(args):
+    # Initialize set of column names
+    # These are reserved names
+    if df_provided:
+        reserved_names = _initialize_argument_col_names(
```
consistency-wise, maybe we should rename this function `_get_reserved_col_names` or something?

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,215 @@ def apply_default_cascade(args):
+    (pandas series type).
+    """
+    df = args["data_frame"]
+    used_col_names = set()
```
here we call it `used_col_names` but elsewhere it's `reserved_names`

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,215 @@ def apply_default_cascade(args):
         args["marginal_x"] = None
 
 
+def _name_heuristic(argument, field_name, used_col_names):
```
here we call it `used_col_names` but elsewhere it's `reserved_names`

emmanuelle replied:
right, there were a few iterations, hence the different names. I will change this.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,215 @@ def apply_default_cascade(args):
+                col_name = argument
+                if isinstance(argument, int):
+                    col_name = _name_heuristic(argument, field, reserved_names)
+                    if field_name not in array_attrables:
```
could we remove this `if/else` block as well as the `continue` and let it "fall through" to the bottom of the loop where we have the same logic?

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -813,6 +813,12 @@ def build_dataframe(args, attrables, array_attrables):
     array_attrables : list
         argument names corresponding to iterables, such as `hover_data`, ...
     """
+    args = dict(input_args)
```
I don't think this makes a deep copy...

```
>>> x= [1,2,3]
>>> y = dict(a=x)
>>> z = dict(y)
>>> z["a"] is x
True
>>> z["a"][0]=100
>>> z
{'a': [100, 2, 3]}
>>> x
[100, 2, 3]
```

But also think this is only an issue for `array_attrables` where we write into `args[something][something]` so if we just copy those that's enough:

```
>>> x = [1,2,3]
>>> y = list(x)
>>> y is x
False
>>> y[0] = 100
>>> x
[1, 2, 3]
```

emmanuelle replied:
oh no! I knew it does a deep copy for a list so I had wrongly assumed I could transpose this pattern to a `dict`.

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -813,6 +813,12 @@ def build_dataframe(args, attrables, array_attrables):
+        if field in array_attrables and isinstance(
+            args[field], pd.core.indexes.base.Index
+        ):
+            args[field] = list(args[field])
```
following from comment above... if we just always do this, instead of in an `if` then I think we're safe (for now!)

emmanuelle replied:
good idea! Then we don't have to do a deepcopy of the whole `args`, it's better not to copy the `dataframe` in particular.

emmanuelle commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,214 @@ def apply_default_cascade(args):
+                continue
+            if isinstance(arg, str):
+                reserved_names.add(arg)
+            if isinstance(arg, pd.core.series.Series):
```
ok thanks, I got the names by asking `type(df)` but more readable names are better!

nicolaskruchten commented:
> What is the simplest way to explain when the .name property of an input series is kept vs. when the arg name is used?

I believe that with what we landed on, the simplest explanation is that `.name` is used only to check if the input is the same object as the correspondingly-named column in `data_frame` (if `data_frame` is provided) otherwise it's ignored. Doing more than that caused all sorts of weird(er) conflicts!

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,214 @@ def apply_default_cascade(args):
+        for arg in names:
+            if arg is None:
+                continue
+            if isinstance(arg, str):
```
In general I agree except for cases where the first `if` results in a `continue` or errors out ;)

emmanuelle commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,214 @@ def apply_default_cascade(args):
+        for arg in names:
+            if arg is None:
+                continue
+            if isinstance(arg, str):
```
not really... but this list of reserved names is tested against the names of keyword arguments which will never be ints (I think) so it should be ok. On the other hand it makes no harm to add int names to the list...

emmanuelle commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,214 @@ def apply_default_cascade(args):
+                    )
+                col_name = argument
+                if isinstance(argument, int):
+                    col_name = _name_heuristic(argument, field, reserved_names)
```
turns out we could write simpler code here. Thanks !

emmanuelle commented:
Thanks a lot for your review @jonmmease, we have iterated a lot on this piece of code and having a fresh eye is wonderful. I think I addressed all your comments, which helped to improve the readability of the code.
Regarding when to use `name`, as explained by @nicolaskruchten we only used names of series equal to columns of the `dataframe` argument. With the possibility to pass arguments coming from different dataframes, handling conflicts was very tricky and hard to explain. Another possibility would have been to take the name whenever it exists and raise an error when a conflict arises, but we preferred to have a strict approach in the beginning and it is still possible in future versions to see if we want to change this behaviour.

nicolaskruchten commented:
Let's do a pass where we ensure all error messages have correct spacing between words: 

`ValueError: All arguments should have the same length.The length of argument `color` is 3, whereas thelength of previous arguments ['x', 'y', 'facet_row', 'facet_col'] is 4` ... "length.The length" and "thelength".

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -754,6 +754,210 @@ def apply_default_cascade(args):
+                            keep_name = df_provided and argument is df_input.index
+                        else:
+                            # we use getattr/hasattr because of index
+                            keep_name = hasattr(
```
```python
df = px.data.tips()
px.scatter(df, x=df["size"], y=df.tip)
```

This has a weird behaviour: it doesn't pick up the name "size" even though it is a valid column... This is almost certainly because `df.size` doesn't return `df["size"]`. Thanks Pandas, again :(

nicolaskruchten replied:
I think the solution is to look up columns via `df_input["name"]` and not `getattr` and just special-case the `index` thing.

emmanuelle replied:
actually it was already a special case, so the modification was easy

nicolaskruchten commented:
OK, other than the spaces in the error messages and the `df.size` thing I think we're good! and we finally have the start of a decent PX test suite: yay!

nicolaskruchten commented on packages/python/plotly/plotly/express/_core.py:
```diff
@@ -929,9 +929,9 @@ def build_dataframe(args, attrables, array_attrables):
-                            keep_name = hasattr(
-                                df_input, col_name
-                            ) and argument is getattr(df_input, col_name)
+                            keep_name = col_name in df_input and argument.equals(
```
Is it intentional to move from `is` to `.equals()` here?

nicolaskruchten replied:
(also we should remove the comment above now ;)

emmanuelle replied:
yes it was intentional, I think it's enough that the values are all equal. What do you think?

emmanuelle replied:
ok I reverted to `is`

nicolaskruchten commented:
💃 💃 💃 thanks so much for getting this thing over the finish line!

emmanuelle commented:
yeehaw ! Merging :-)

thatneat commented:
Would love to consume this from PyPI! Is there a release planned?

nicolaskruchten commented:
Version 4.2 should come out this week, yes!
